Option to pause Polling when window Unfocused

Sorry if this is a silly question, but is it possible to pause polling on a query if the window loses focus.

I currently have a situation where i need to manually `initiate` a query due to some other bad code where i have used prefetch in a horrible hacky way to bypass cache timeout (it's a pagination/inifinate scroll issue that is first on my list for a refactor!!).

```typescript
  useEffect(() => {
    let unsubscribe: () => void;

    if (id) {
      const result = dispatch(
        getChatHistory.initiate(
          { id, page: 1 },
          {
            subscribe: true,
            subscriptionOptions: {
              pollingInterval: 10000,
            },
          }
        )
      );

      unsubscribe = result.unsubscribe;
    }

    return () => {
      unsubscribe && unsubscribe();
    };
  }, [id, dispatch]);
```

If there is no "native" way to pause polling, I think it should be possible to create a listener for the focus/unfocus action dispatches and set some state value accordingly.

I have posted here instead of SO in case the functionality could be useful (please close/delete the issue if there is no merit in the idea and I will ask on SO instead).

I would love to be able to contribute in some way, but this one might be out of my wheelhouse. I have solved my issue in the short term by creating actions for the Focus events and then setting state in `extraReducers`.

```typescript
const onFocus = createAction("__rtkq/focused");
const onFocusLost = createAction("__rtkq/unfocused");
```

I see these actions are exported here:

https://github.com/reduxjs/redux-toolkit/blob/adf7d01e350daa1f6480eab86a19424ef6a5b6a1/packages/toolkit/src/query/core/setupListeners.ts#L7

But I was unable to import them and use them in my project (NextJS).

VSCode is happy to import them like this

```typescript
import {
  onFocus,
  onFocusLost,
} from "@reduxjs/toolkit/dist/query/core/setupListeners";
```

Typescript is happy with it but NextJS isn't. Running `dev` or `build` results in

```
Module not found: Can't resolve '@reduxjs/toolkit/dist/query/core/setupListeners' in 'C:\Code\project\store\features\messages'
```

While creating my own actions has worked, i'm not a fan of the idea long term as it depends on RTK not changing the ActionType at some point in the future.

Is it possible to get these Actions exported so they can be used by projects as needed?

Now this is something that I could create a PR for if you agree it is a worthy exercise :-).

Thanks for your help.

--- a/packages/toolkit/src/query/core/apiState.ts
+++ b/packages/toolkit/src/query/core/apiState.ts
@@ -87,6 +87,8 @@ export type SubscriptionOptions = {
    *  Defaults to 'false'. This setting allows you to control whether RTK Query will continue polling if the window is not focused.
    *
    *  If pollingInterval is not set or set to 0, this **will not be evaluated** until pollingInterval is greater than 0.
+   *
+   *  Note: requires [`setupListeners`](./setupListeners) to have been called.
    */
   skipPollOnFocusLost?: boolean
   /**

--- a/packages/toolkit/src/query/core/apiState.ts
+++ b/packages/toolkit/src/query/core/apiState.ts
@@ -84,6 +84,12 @@ export type SubscriptionOptions = {
    */
   pollingInterval?: number
   /**
+   *  Defaults to 'false'. This setting allows you to control whether RTK Query will continue polling if the window is not focused.
+   *
+   *  If pollingInterval is not set or set to 0, this **will not be evaluated** until pollingInterval is greater than 0.
+   */
+  skipPollOnFocusLost?: boolean
+  /**
    * Defaults to `false`. This setting allows you to control whether RTK Query will try to refetch all subscribed queries after regaining a network connection.
    *
    * If you specify this option alongside `skip: true`, this **will not be evaluated** until `skip` is false.

--- a/packages/toolkit/src/query/core/apiState.ts
+++ b/packages/toolkit/src/query/core/apiState.ts
@@ -90,7 +90,7 @@ export type SubscriptionOptions = {
    *
    *  Note: requires [`setupListeners`](./setupListeners) to have been called.
    */
-  skipPollOnFocusLost?: boolean
+  skipPollingIfUnfocused?: boolean
   /**
    * Defaults to `false`. This setting allows you to control whether RTK Query will try to refetch all subscribed queries after regaining a network connection.
    *

--- a/packages/toolkit/src/query/core/buildMiddleware/polling.ts
+++ b/packages/toolkit/src/query/core/buildMiddleware/polling.ts
@@ -53,38 +53,36 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
     { queryCacheKey }: QuerySubstateIdentifier,
     api: SubMiddlewareApi
   ) {
-    const state = api.getState()[reducerPath];
-    const querySubState = state.queries[queryCacheKey];
-    const subscriptions = internalState.currentSubscriptions[queryCacheKey];
+    const state = api.getState()[reducerPath]
+    const querySubState = state.queries[queryCacheKey]
+    const subscriptions = internalState.currentSubscriptions[queryCacheKey]
 
     if (!querySubState || querySubState.status === QueryStatus.uninitialized)
-      return;
+      return
 
-    const { lowestPollingInterval, skipPollOnFocusLost } = findLowestPollingInterval(subscriptions);
-    if (!Number.isFinite(lowestPollingInterval)) return;
+    const { lowestPollingInterval, skipPollOnFocusLost } =
+      findLowestPollingInterval(subscriptions)
+    if (!Number.isFinite(lowestPollingInterval)) return
 
-    const currentPoll = currentPolls[queryCacheKey];
+    const currentPoll = currentPolls[queryCacheKey]
 
     if (currentPoll?.timeout) {
-      clearTimeout(currentPoll.timeout);
-      currentPoll.timeout = undefined;
+      clearTimeout(currentPoll.timeout)
+      currentPoll.timeout = undefined
     }
 
-    const nextPollTimestamp = Date.now() + lowestPollingInterval;
+    const nextPollTimestamp = Date.now() + lowestPollingInterval
 
-    // Always update the polling interval
     currentPolls[queryCacheKey] = {
       nextPollTimestamp,
       pollingInterval: lowestPollingInterval,
       timeout: setTimeout(() => {
-        // Conditionally dispatch the query
         if (document.hasFocus() || !skipPollOnFocusLost) {
-          api.dispatch(refetchQuery(querySubState, queryCacheKey));
+          api.dispatch(refetchQuery(querySubState, queryCacheKey))
         }
-        // Regardless of dispatch, set up the next poll
-        startNextPoll({ queryCacheKey }, api);
+        startNextPoll({ queryCacheKey }, api)
       }, lowestPollingInterval),
-    };
+    }
   }
 
   function updatePollingInterval(

--- a/packages/toolkit/src/query/core/buildMiddleware/polling.ts
+++ b/packages/toolkit/src/query/core/buildMiddleware/polling.ts
@@ -53,35 +53,38 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
     { queryCacheKey }: QuerySubstateIdentifier,
     api: SubMiddlewareApi
   ) {
-    const state = api.getState()[reducerPath]
-    const querySubState = state.queries[queryCacheKey]
-    const subscriptions = internalState.currentSubscriptions[queryCacheKey]
+    const state = api.getState()[reducerPath];
+    const querySubState = state.queries[queryCacheKey];
+    const subscriptions = internalState.currentSubscriptions[queryCacheKey];
 
     if (!querySubState || querySubState.status === QueryStatus.uninitialized)
-      return
+      return;
 
-    const lowestPollingInterval = findLowestPollingInterval(subscriptions)
-    if (!Number.isFinite(lowestPollingInterval)) return
+    const { lowestPollingInterval, skipPollOnFocusLost } = findLowestPollingInterval(subscriptions);
+    if (!Number.isFinite(lowestPollingInterval)) return;
 
-    const currentPoll = currentPolls[queryCacheKey]
+    const currentPoll = currentPolls[queryCacheKey];
 
     if (currentPoll?.timeout) {
-      clearTimeout(currentPoll.timeout)
-      currentPoll.timeout = undefined
+      clearTimeout(currentPoll.timeout);
+      currentPoll.timeout = undefined;
     }
 
-    const nextPollTimestamp = Date.now() + lowestPollingInterval
+    const nextPollTimestamp = Date.now() + lowestPollingInterval;
 
-    const currentInterval: typeof currentPolls[number] = (currentPolls[
-      queryCacheKey
-    ] = {
+    // Always update the polling interval
+    currentPolls[queryCacheKey] = {
       nextPollTimestamp,
       pollingInterval: lowestPollingInterval,
       timeout: setTimeout(() => {
-        currentInterval!.timeout = undefined
-        api.dispatch(refetchQuery(querySubState, queryCacheKey))
+        // Conditionally dispatch the query
+        if (document.hasFocus() || !skipPollOnFocusLost) {
+          api.dispatch(refetchQuery(querySubState, queryCacheKey));
+        }
+        // Regardless of dispatch, set up the next poll
+        startNextPoll({ queryCacheKey }, api);
       }, lowestPollingInterval),
-    })
+    };
   }
 
   function updatePollingInterval(
@@ -96,7 +99,7 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
       return
     }
 
-    const lowestPollingInterval = findLowestPollingInterval(subscriptions)
+    const { lowestPollingInterval } = findLowestPollingInterval(subscriptions)
 
     if (!Number.isFinite(lowestPollingInterval)) {
       cleanupPollForKey(queryCacheKey)
@@ -126,6 +129,7 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
   }
 
   function findLowestPollingInterval(subscribers: Subscribers = {}) {
+    let skipPollOnFocusLost: boolean | undefined = false
     let lowestPollingInterval = Number.POSITIVE_INFINITY
     for (let key in subscribers) {
       if (!!subscribers[key].pollingInterval) {
@@ -133,10 +137,15 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
           subscribers[key].pollingInterval!,
           lowestPollingInterval
         )
+        skipPollOnFocusLost = subscribers[key].skipPollOnFocusLost
       }
     }
 
-    return lowestPollingInterval
+    return {
+      lowestPollingInterval,
+      skipPollOnFocusLost,
+    }
   }
+
   return handler
 }

--- a/packages/toolkit/src/query/core/buildMiddleware/polling.ts
+++ b/packages/toolkit/src/query/core/buildMiddleware/polling.ts
@@ -60,7 +60,7 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
     if (!querySubState || querySubState.status === QueryStatus.uninitialized)
       return
 
-    const { lowestPollingInterval, skipPollOnFocusLost } =
+    const { lowestPollingInterval, skipPollingIfUnfocused } =
       findLowestPollingInterval(subscriptions)
     if (!Number.isFinite(lowestPollingInterval)) return
 
@@ -77,7 +77,7 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
       nextPollTimestamp,
       pollingInterval: lowestPollingInterval,
       timeout: setTimeout(() => {
-        if (state.config.focused || !skipPollOnFocusLost) {
+        if (state.config.focused || !skipPollingIfUnfocused) {
           api.dispatch(refetchQuery(querySubState, queryCacheKey))
         }
         startNextPoll({ queryCacheKey }, api)
@@ -127,7 +127,7 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
   }
 
   function findLowestPollingInterval(subscribers: Subscribers = {}) {
-    let skipPollOnFocusLost: boolean | undefined = false
+    let skipPollingIfUnfocused: boolean | undefined = false
     let lowestPollingInterval = Number.POSITIVE_INFINITY
     for (let key in subscribers) {
       if (!!subscribers[key].pollingInterval) {
@@ -135,14 +135,14 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
           subscribers[key].pollingInterval!,
           lowestPollingInterval
         )
-        skipPollOnFocusLost =
-          subscribers[key].skipPollOnFocusLost || skipPollOnFocusLost
+        skipPollingIfUnfocused =
+          subscribers[key].skipPollingIfUnfocused || skipPollingIfUnfocused
       }
     }
 
     return {
       lowestPollingInterval,
-      skipPollOnFocusLost,
+      skipPollingIfUnfocused,
     }
   }
 

--- a/packages/toolkit/src/query/core/buildMiddleware/polling.ts
+++ b/packages/toolkit/src/query/core/buildMiddleware/polling.ts
@@ -135,9 +135,7 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
           subscribers[key].pollingInterval!,
           lowestPollingInterval
         )
-        // if (!skipPollOnFocusLost) {
-          skipPollOnFocusLost = subscribers[key].skipPollOnFocusLost
-        // }
+        skipPollOnFocusLost = subscribers[key].skipPollOnFocusLost
       }
     }
 

--- a/packages/toolkit/src/query/core/buildMiddleware/polling.ts
+++ b/packages/toolkit/src/query/core/buildMiddleware/polling.ts
@@ -135,7 +135,8 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
           subscribers[key].pollingInterval!,
           lowestPollingInterval
         )
-        skipPollOnFocusLost = subscribers[key].skipPollOnFocusLost
+        skipPollOnFocusLost =
+          subscribers[key].skipPollOnFocusLost || skipPollOnFocusLost
       }
     }
 

--- a/packages/toolkit/src/query/core/buildMiddleware/polling.ts
+++ b/packages/toolkit/src/query/core/buildMiddleware/polling.ts
@@ -77,7 +77,7 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
       nextPollTimestamp,
       pollingInterval: lowestPollingInterval,
       timeout: setTimeout(() => {
-        if (document.hasFocus() || !skipPollOnFocusLost) {
+        if (state.config.focused || !skipPollOnFocusLost) {
           api.dispatch(refetchQuery(querySubState, queryCacheKey))
         }
         startNextPoll({ queryCacheKey }, api)
@@ -135,7 +135,9 @@ export const buildPollingHandler: InternalHandlerBuilder = ({
           subscribers[key].pollingInterval!,
           lowestPollingInterval
         )
-        skipPollOnFocusLost = subscribers[key].skipPollOnFocusLost
+        // if (!skipPollOnFocusLost) {
+          skipPollOnFocusLost = subscribers[key].skipPollOnFocusLost
+        // }
       }
     }
 

--- a/packages/toolkit/src/query/react/buildHooks.ts
+++ b/packages/toolkit/src/query/react/buildHooks.ts
@@ -672,7 +672,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
         refetchOnMountOrArgChange,
         skip = false,
         pollingInterval = 0,
-        skipPollOnFocusLost = false,
+        skipPollingIfUnfocused = false,
       } = {}
     ) => {
       const { initiate } = api.endpoints[name] as ApiEndpointQuery<
@@ -716,7 +716,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
         refetchOnReconnect,
         refetchOnFocus,
         pollingInterval,
-        skipPollOnFocusLost,
+        skipPollingIfUnfocused,
       })
 
       const lastRenderHadSubscription = useRef(false)
@@ -817,7 +817,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
       refetchOnReconnect,
       refetchOnFocus,
       pollingInterval = 0,
-      skipPollOnFocusLost = false,
+      skipPollingIfUnfocused = false,
     } = {}) => {
       const { initiate } = api.endpoints[name] as ApiEndpointQuery<
         QueryDefinition<any, any, any, any, any>,
@@ -832,7 +832,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
         refetchOnReconnect,
         refetchOnFocus,
         pollingInterval,
-        skipPollOnFocusLost,
+        skipPollingIfUnfocused,
       })
 
       usePossiblyImmediateEffect(() => {

--- a/packages/toolkit/src/query/react/buildHooks.ts
+++ b/packages/toolkit/src/query/react/buildHooks.ts
@@ -672,6 +672,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
         refetchOnMountOrArgChange,
         skip = false,
         pollingInterval = 0,
+        skipPollOnFocusLost = false,
       } = {}
     ) => {
       const { initiate } = api.endpoints[name] as ApiEndpointQuery<
@@ -715,6 +716,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
         refetchOnReconnect,
         refetchOnFocus,
         pollingInterval,
+        skipPollOnFocusLost,
       })
 
       const lastRenderHadSubscription = useRef(false)
@@ -815,6 +817,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
       refetchOnReconnect,
       refetchOnFocus,
       pollingInterval = 0,
+      skipPollOnFocusLost = false,
     } = {}) => {
       const { initiate } = api.endpoints[name] as ApiEndpointQuery<
         QueryDefinition<any, any, any, any, any>,
@@ -829,6 +832,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
         refetchOnReconnect,
         refetchOnFocus,
         pollingInterval,
+        skipPollOnFocusLost,
       })
 
       usePossiblyImmediateEffect(() => {

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -122,4 +122,85 @@ describe('polling tests', () => {
 
     expect(mockBaseQuery.mock.calls.length).toBeGreaterThanOrEqual(2)
   })
+
+  it('respects skipPollOnFocusLost', async () => {
+    storeRef.store.dispatch(
+      getPosts.initiate(1, {
+        subscriptionOptions: { pollingInterval: 10, skipPollOnFocusLost: true },
+        subscribe: true,
+      })
+    )
+
+    await delay(20)
+    const callsWithSkip = mockBaseQuery.mock.calls.length
+
+    storeRef.store.dispatch(
+      getPosts.initiate(1, {
+        subscriptionOptions: {
+          pollingInterval: 10,
+          skipPollOnFocusLost: false,
+        },
+        subscribe: true,
+      })
+    )
+
+    await delay(30)
+    const callsWithoutSkip = mockBaseQuery.mock.calls.length
+
+    expect(callsWithSkip).toBe(1)
+    expect(callsWithoutSkip).toBeGreaterThan(2)
+  })
+
+  it('replaces skipPollOnFocusLost with most recent mount', async () => {
+    storeRef.store.dispatch(
+      getPosts.initiate(1, {
+        subscriptionOptions: {
+          pollingInterval: 10,
+          skipPollOnFocusLost: false,
+        },
+        subscribe: true,
+      })
+    )
+
+    await delay(50)
+    const callsWithSkip = mockBaseQuery.mock.calls.length
+
+    storeRef.store.dispatch(
+      getPosts.initiate(1, {
+        subscriptionOptions: { pollingInterval: 15, skipPollOnFocusLost: true },
+        subscribe: true,
+      })
+    )
+
+    await delay(50)
+    const callsWithoutSkip = mockBaseQuery.mock.calls.length
+
+    expect(callsWithSkip).toBeGreaterThan(2)
+    expect(callsWithoutSkip).toBe(callsWithSkip + 1)
+  })
+
+  it('replaces skipPollOnFocusLost when the subscription options are updated', async () => {
+    const { requestId, queryCacheKey, ...subscription } =
+      storeRef.store.dispatch(
+        getPosts.initiate(1, {
+          subscriptionOptions: { pollingInterval: 10 },
+          subscribe: true,
+        })
+      )
+
+    const getSubs = createSubscriptionGetter(queryCacheKey)
+
+    await delay(1)
+    expect(Object.keys(getSubs())).toHaveLength(1)
+    expect(getSubs()[requestId].skipPollOnFocusLost).toBe(false)
+
+    subscription.updateSubscriptionOptions({
+      pollingInterval: 20,
+      skipPollOnFocusLost: true,
+    })
+
+    await delay(1)
+    expect(Object.keys(getSubs())).toHaveLength(1)
+    expect(getSubs()[requestId].skipPollOnFocusLost).toBe(true)
+  })
 })

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -123,11 +123,14 @@ describe('polling tests', () => {
     expect(mockBaseQuery.mock.calls.length).toBeGreaterThanOrEqual(2)
   })
 
-  it('respects skipPollOnFocusLost', async () => {
+  it('respects skipPollingIfUnfocused', async () => {
     mockBaseQuery.mockClear()
     storeRef.store.dispatch(
       getPosts.initiate(2, {
-        subscriptionOptions: { pollingInterval: 10, skipPollOnFocusLost: true },
+        subscriptionOptions: {
+          pollingInterval: 10,
+          skipPollingIfUnfocused: true,
+        },
         subscribe: true,
       })
     )
@@ -140,7 +143,7 @@ describe('polling tests', () => {
       getPosts.initiate(2, {
         subscriptionOptions: {
           pollingInterval: 10,
-          skipPollOnFocusLost: false,
+          skipPollingIfUnfocused: false,
         },
         subscribe: true,
       })
@@ -157,12 +160,12 @@ describe('polling tests', () => {
     storeRef.store.dispatch(api.util.resetApiState())
   })
 
-  it('respects skipPollOnFocusLost if at least one subscription has it', async () => {
+  it('respects skipPollingIfUnfocused if at least one subscription has it', async () => {
     storeRef.store.dispatch(
       getPosts.initiate(3, {
         subscriptionOptions: {
           pollingInterval: 10,
-          skipPollOnFocusLost: false,
+          skipPollingIfUnfocused: false,
         },
         subscribe: true,
       })
@@ -173,7 +176,10 @@ describe('polling tests', () => {
 
     storeRef.store.dispatch(
       getPosts.initiate(3, {
-        subscriptionOptions: { pollingInterval: 15, skipPollOnFocusLost: true },
+        subscriptionOptions: {
+          pollingInterval: 15,
+          skipPollingIfUnfocused: true,
+        },
         subscribe: true,
       })
     )
@@ -182,7 +188,7 @@ describe('polling tests', () => {
       getPosts.initiate(3, {
         subscriptionOptions: {
           pollingInterval: 20,
-          skipPollOnFocusLost: false,
+          skipPollingIfUnfocused: false,
         },
         subscribe: true,
       })
@@ -197,13 +203,13 @@ describe('polling tests', () => {
     expect(callsWithSkip).toBe(callsWithoutSkip + 1)
   })
 
-  it('replaces skipPollOnFocusLost when the subscription options are updated', async () => {
+  it('replaces skipPollingIfUnfocused when the subscription options are updated', async () => {
     const { requestId, queryCacheKey, ...subscription } =
       storeRef.store.dispatch(
         getPosts.initiate(1, {
           subscriptionOptions: {
             pollingInterval: 10,
-            skipPollOnFocusLost: false,
+            skipPollingIfUnfocused: false,
           },
           subscribe: true,
         })
@@ -213,15 +219,15 @@ describe('polling tests', () => {
 
     await delay(1)
     expect(Object.keys(getSubs())).toHaveLength(1)
-    expect(getSubs()[requestId].skipPollOnFocusLost).toBe(false)
+    expect(getSubs()[requestId].skipPollingIfUnfocused).toBe(false)
 
     subscription.updateSubscriptionOptions({
       pollingInterval: 20,
-      skipPollOnFocusLost: true,
+      skipPollingIfUnfocused: true,
     })
 
     await delay(1)
     expect(Object.keys(getSubs())).toHaveLength(1)
-    expect(getSubs()[requestId].skipPollOnFocusLost).toBe(true)
+    expect(getSubs()[requestId].skipPollingIfUnfocused).toBe(true)
   })
 })

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -2,6 +2,8 @@ import { createApi } from '@reduxjs/toolkit/query'
 import { delay } from 'msw'
 import { setupApiStore } from './helpers'
 import type { SubscriptionSelectors } from '../core/buildMiddleware/types'
+import { createListenerMiddleware } from '@reduxjs/toolkit'
+
 
 const mockBaseQuery = vi
   .fn()
@@ -125,18 +127,27 @@ describe('polling tests', () => {
 
   it('respects skipPollOnFocusLost', async () => {
     mockBaseQuery.mockClear()
-    storeRef.store.dispatch(
-      getPosts.initiate(1, {
+    const listenerMiddleware = createListenerMiddleware()
+    const storeListenerRef = setupApiStore(api, undefined, {
+      middleware: {
+        concat: [listenerMiddleware.middleware],
+      },
+      withoutTestLifecycles: true,
+    })
+
+    storeListenerRef.store.dispatch(
+      getPosts.initiate(2, {
         subscriptionOptions: { pollingInterval: 10, skipPollOnFocusLost: true },
         subscribe: true,
       })
     )
+    storeListenerRef.store.dispatch(api.internalActions?.onFocusLost())
 
-    await delay(20)
+    await delay(50)
     const callsWithSkip = mockBaseQuery.mock.calls.length
 
-    storeRef.store.dispatch(
-      getPosts.initiate(1, {
+    storeListenerRef.store.dispatch(
+      getPosts.initiate(2, {
         subscriptionOptions: {
           pollingInterval: 10,
           skipPollOnFocusLost: false,
@@ -145,17 +156,28 @@ describe('polling tests', () => {
       })
     )
 
-    await delay(30)
+    storeListenerRef.store.dispatch(api.internalActions?.onFocus())
+
+    await delay(50)
     const callsWithoutSkip = mockBaseQuery.mock.calls.length
-    console.log(callsWithSkip, callsWithoutSkip)
 
     expect(callsWithSkip).toBe(1)
     expect(callsWithoutSkip).toBeGreaterThan(2)
+
+    storeListenerRef.store.dispatch(api.util.resetApiState())
   })
 
-  it('replaces skipPollOnFocusLost with most recent mount', async () => {
-    storeRef.store.dispatch(
-      getPosts.initiate(1, {
+  it('respects skipPollOnFocusLost if any subscription is true', async () => {
+    const listenerMiddleware = createListenerMiddleware()
+    const storeListenerRef = setupApiStore(api, undefined, {
+      middleware: {
+        concat: [listenerMiddleware.middleware],
+      },
+      withoutTestLifecycles: true,
+    })
+
+    storeListenerRef.store.dispatch(
+      getPosts.initiate(3, {
         subscriptionOptions: {
           pollingInterval: 10,
           skipPollOnFocusLost: false,
@@ -167,12 +189,14 @@ describe('polling tests', () => {
     await delay(50)
     const callsWithSkip = mockBaseQuery.mock.calls.length
 
-    storeRef.store.dispatch(
-      getPosts.initiate(1, {
+    storeListenerRef.store.dispatch(
+      getPosts.initiate(3, {
         subscriptionOptions: { pollingInterval: 15, skipPollOnFocusLost: true },
         subscribe: true,
       })
     )
+
+    storeListenerRef.store.dispatch(api.internalActions?.onFocusLost())
 
     await delay(50)
     const callsWithoutSkip = mockBaseQuery.mock.calls.length

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -2,7 +2,6 @@ import { createApi } from '@reduxjs/toolkit/query'
 import { delay } from 'msw'
 import { setupApiStore } from './helpers'
 import type { SubscriptionSelectors } from '../core/buildMiddleware/types'
-import { createListenerMiddleware } from '@reduxjs/toolkit'
 
 const mockBaseQuery = vi
   .fn()
@@ -126,26 +125,18 @@ describe('polling tests', () => {
 
   it('respects skipPollOnFocusLost', async () => {
     mockBaseQuery.mockClear()
-    const listenerMiddleware = createListenerMiddleware()
-    const storeListenerRef = setupApiStore(api, undefined, {
-      middleware: {
-        concat: [listenerMiddleware.middleware],
-      },
-      withoutTestLifecycles: true,
-    })
-
-    storeListenerRef.store.dispatch(
+    storeRef.store.dispatch(
       getPosts.initiate(2, {
         subscriptionOptions: { pollingInterval: 10, skipPollOnFocusLost: true },
         subscribe: true,
       })
     )
-    storeListenerRef.store.dispatch(api.internalActions?.onFocusLost())
+    storeRef.store.dispatch(api.internalActions?.onFocusLost())
 
     await delay(50)
     const callsWithSkip = mockBaseQuery.mock.calls.length
 
-    storeListenerRef.store.dispatch(
+    storeRef.store.dispatch(
       getPosts.initiate(2, {
         subscriptionOptions: {
           pollingInterval: 10,
@@ -155,7 +146,7 @@ describe('polling tests', () => {
       })
     )
 
-    storeListenerRef.store.dispatch(api.internalActions?.onFocus())
+    storeRef.store.dispatch(api.internalActions?.onFocus())
 
     await delay(50)
     const callsWithoutSkip = mockBaseQuery.mock.calls.length
@@ -163,19 +154,11 @@ describe('polling tests', () => {
     expect(callsWithSkip).toBe(1)
     expect(callsWithoutSkip).toBeGreaterThan(2)
 
-    storeListenerRef.store.dispatch(api.util.resetApiState())
+    storeRef.store.dispatch(api.util.resetApiState())
   })
 
   it('respects skipPollOnFocusLost of the most recent mounted subscription', async () => {
-    const listenerMiddleware = createListenerMiddleware()
-    const storeListenerRef = setupApiStore(api, undefined, {
-      middleware: {
-        concat: [listenerMiddleware.middleware],
-      },
-      withoutTestLifecycles: true,
-    })
-
-    storeListenerRef.store.dispatch(
+    storeRef.store.dispatch(
       getPosts.initiate(3, {
         subscriptionOptions: {
           pollingInterval: 10,
@@ -188,14 +171,14 @@ describe('polling tests', () => {
     await delay(50)
     const callsWithSkip = mockBaseQuery.mock.calls.length
 
-    storeListenerRef.store.dispatch(
+    storeRef.store.dispatch(
       getPosts.initiate(3, {
         subscriptionOptions: { pollingInterval: 15, skipPollOnFocusLost: true },
         subscribe: true,
       })
     )
 
-    storeListenerRef.store.dispatch(api.internalActions?.onFocusLost())
+    storeRef.store.dispatch(api.internalActions?.onFocusLost())
 
     await delay(50)
     const callsWithoutSkip = mockBaseQuery.mock.calls.length

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -4,7 +4,6 @@ import { setupApiStore } from './helpers'
 import type { SubscriptionSelectors } from '../core/buildMiddleware/types'
 import { createListenerMiddleware } from '@reduxjs/toolkit'
 
-
 const mockBaseQuery = vi
   .fn()
   .mockImplementation((args: any) => ({ data: args }))
@@ -167,7 +166,7 @@ describe('polling tests', () => {
     storeListenerRef.store.dispatch(api.util.resetApiState())
   })
 
-  it('respects skipPollOnFocusLost if any subscription is true', async () => {
+  it('respects skipPollOnFocusLost of the most recent mounted subscription', async () => {
     const listenerMiddleware = createListenerMiddleware()
     const storeListenerRef = setupApiStore(api, undefined, {
       middleware: {

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -185,7 +185,10 @@ describe('polling tests', () => {
     const { requestId, queryCacheKey, ...subscription } =
       storeRef.store.dispatch(
         getPosts.initiate(1, {
-          subscriptionOptions: { pollingInterval: 10, skipPollOnFocusLost: false },
+          subscriptionOptions: {
+            pollingInterval: 10,
+            skipPollOnFocusLost: false,
+          },
           subscribe: true,
         })
       )

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -124,6 +124,7 @@ describe('polling tests', () => {
   })
 
   it('respects skipPollOnFocusLost', async () => {
+    mockBaseQuery.mockClear()
     storeRef.store.dispatch(
       getPosts.initiate(1, {
         subscriptionOptions: { pollingInterval: 10, skipPollOnFocusLost: true },
@@ -146,6 +147,7 @@ describe('polling tests', () => {
 
     await delay(30)
     const callsWithoutSkip = mockBaseQuery.mock.calls.length
+    console.log(callsWithSkip, callsWithoutSkip)
 
     expect(callsWithSkip).toBe(1)
     expect(callsWithoutSkip).toBeGreaterThan(2)
@@ -183,7 +185,7 @@ describe('polling tests', () => {
     const { requestId, queryCacheKey, ...subscription } =
       storeRef.store.dispatch(
         getPosts.initiate(1, {
-          subscriptionOptions: { pollingInterval: 10 },
+          subscriptionOptions: { pollingInterval: 10, skipPollOnFocusLost: false },
           subscribe: true,
         })
       )

--- a/packages/toolkit/src/query/tests/polling.test.tsx
+++ b/packages/toolkit/src/query/tests/polling.test.tsx
@@ -157,7 +157,7 @@ describe('polling tests', () => {
     storeRef.store.dispatch(api.util.resetApiState())
   })
 
-  it('respects skipPollOnFocusLost of the most recent mounted subscription', async () => {
+  it('respects skipPollOnFocusLost if at least one subscription has it', async () => {
     storeRef.store.dispatch(
       getPosts.initiate(3, {
         subscriptionOptions: {
@@ -169,7 +169,7 @@ describe('polling tests', () => {
     )
 
     await delay(50)
-    const callsWithSkip = mockBaseQuery.mock.calls.length
+    const callsWithoutSkip = mockBaseQuery.mock.calls.length
 
     storeRef.store.dispatch(
       getPosts.initiate(3, {
@@ -178,13 +178,23 @@ describe('polling tests', () => {
       })
     )
 
+    storeRef.store.dispatch(
+      getPosts.initiate(3, {
+        subscriptionOptions: {
+          pollingInterval: 20,
+          skipPollOnFocusLost: false,
+        },
+        subscribe: true,
+      })
+    )
+
     storeRef.store.dispatch(api.internalActions?.onFocusLost())
 
     await delay(50)
-    const callsWithoutSkip = mockBaseQuery.mock.calls.length
+    const callsWithSkip = mockBaseQuery.mock.calls.length
 
-    expect(callsWithSkip).toBeGreaterThan(2)
-    expect(callsWithoutSkip).toBe(callsWithSkip + 1)
+    expect(callsWithoutSkip).toBeGreaterThan(2)
+    expect(callsWithSkip).toBe(callsWithoutSkip + 1)
   })
 
   it('replaces skipPollOnFocusLost when the subscription options are updated', async () => {

